Feature/Migrating from 7.0 to 7.1 (#118)
smartdevicelink/sdl_core_guides: docs/Migrating to Newer SDL Versions/Migrating SDL Core 7.0 to 7.1/index.md

# Migrating SDL Core 7.0 to 7.1

The 7.1 release had a number of changes and additions to the HMI API that will require updates to your SDL Core integration in your head unit. 

## Environment Update

> The default supported version was changed to Ubuntu 20. Recommended GCC Version 9.3.x.

Reviewer: ```suggestion
The default supported version was changed to Ubuntu 20. Recommended GCC Version 9.3.x.

Support was added for OpenSSL 1.1, we recommend updating your version of the library accordingly.
Along with support for OpenSSL 1.1, a configurable `SecurityLevel` field was added to the INI file. This value can be customized depending on the security requirements of your system (see the [OpenSSL documentation](https://www.openssl.org/docs/man1.1.0/man3/SSL_CTX_get_security_level.html) for a description of each security level)
```

## Newly Deprecated

### Deprecated SyncPData RPCs

- RPC request and response for `EncodedSyncPData` has been marked as deprecated.
- RPC notification `OnEncodedSyncPData` has been marked as deprecated.

### Deprecated UI params

- `TextFieldName` element `mediaClock` has been marked as deprecated.

> - `Show` RPC param `mediaClock`has been marked as deprecated.

Reviewer: ```suggestion
- `Show` RPC param `mediaClock` has been marked as deprecated.
```

## Additions

### Vehicle Data

- `BodyInformation` was expanded to replace `driverDoorAjar`, `passengerDoorAjar`, `rearLeftDoorAjar` and `rearRightDoorAjar` parameters.
- New vehicle data type: `climateData` to replace `externalTemperature` parameter.
- New vehicle data type: `SeatOccupancy`.

It is not required to implement all vehicle data types. If a type is unsupported by your headunit, please be sure to respond to SDL Core with result `UNSUPPORTED_RESOURCE` if an unsupported request has been made.

### HMI UI Additions

#### Custom Plkayback rates for SetMediaClockTimer

A media app now has the ability to set the media playback timer and bar to advance at a custom rate (like 125% speed).

Added new parameter `countRate` to the `SetMediaClockTimer RPC`

```xml
<function name="SetMediaClockTimer" functionID="SetMediaClockTimerID" messagetype="request" since="1.0">
    <description>Sets the initial media clock value and automatic update method.</description>
        
    <!-- New Parameter -->

>     <param name="countRate" type="Float" minValue="0.1" maxValue="100.0" defvalue="1.0" mandatory="false">

Reviewer: ```suggestion
    <param name="countRate" type="Float" minvalue="0.1" maxvalue="100.0" defvalue="1.0" mandatory="false">
```

        The value of this parameter is the amount that the media clock timer will advance per 1.0 seconds of real time.
        
        Values less than 1.0 will therefore advance the timer slower than real-time, while values greater than 1.0 will advance the timer faster than real-time.

        e.g. If this parameter is set to `0.5`, the timer will advance one second per two seconds real-time, or at 50% speed. If this parameter is set to `2.0`, the timer will advance two seconds per one second real-time, or at 200% speed.
        </description>
    </param>
</function>
```

#### Media Skip Indicators

A media app now has the ability to change the indicators for the `SEEKLEFT` and `SEEKRIGHT` buttons to show either time skip buttons or track skip buttons.

- Added new parameters `forwardSeekIndicator` and `backSeekIndicator` to the `SetMediaClockTimer RPC`.

```xml
<enum name="SeekIndicatorType">
    <element name="TRACK">
    <element name="TIME">
</enum>

<struct name="SeekStreamingIndicator">
    <description>
        The seek next / skip previous subscription buttons' content
    </description> 

    <param name="type" type="SeekIndicatorType" mandatory="true" />

>     <param name="seekTime" type="Int" minValue="1" maxValue="99" mandatory="false">

Reviewer: ```suggestion
    <param name="seekTime" type="Integer" minvalue="1" maxvalue="99" mandatory="false">
```

    <description>
        Adds a command to the in application menu.
        Either menuParams or vrCommands must be provided.
    </description>
    
    <!-- New Parameters -->
    <param name="secondaryImage" type="Image" mandatory="false">
        <description>Optional secondary image struct for menu cell</description>
    </param>
</function>

<struct name="MenuParams" since="1.0">
    <!-- New Parameters -->
    <param name="secondaryText" maxlength="500" type="String" mandatory="false">
        <description>Optional secondary text to display</description>
    </param>
    <param name="tertiaryText" maxlength="500" type="String" mandatory="false">
        <description>Optional tertiary text to display</description>
    </param>
</struct>
```

#### Broadening choice uniqueness

Prior to SDL Core 7.1, choice set choices and menu commands were required to have unique primary text. SDL Core 7.1 removes this restirction 

#### Keyboard Enhancements

SDL Core 7.1 adds a new `NUMERIC` keyboard layout and new enhancements to allow apps to mask entered characters and change special characters shown on the keyboard layout.

> ### OEM exlcusive apps support

Reviewer: ```suggestion
### OEM exclusive apps support
```

|make|String|Vehicle make|
|model|String|Vehicle model|
|model year|String|Vehicle model year|
|trim|String|Vehicle trim|
|systemSoftwareVersion|String|Vehicle system software version|
|systemHardwareVersion|String|Vehicle system hardware version|

The vehicle type information parameters (`vehicleType` and `systemSoftwareVersion`) in `RegisterAppInterface` have been deprecated
### Video streaming capability updates

#### Preffered FPS

- Added new parameter `prefferedFPS` to the `VideoStreamingCapability` struct.

```xml
<struct name="VideoStreamingCapability" since="4.5">
    <description>Contains information about this system's video streaming capabilities.</description>
    ...
    <!-- new param -->
    <param name="preferredFPS" type="Integer" minvalue="0" maxvalue="2147483647" mandatory="false">
        <description>The preferred frame rate per second of the head unit. The mobile application / app library may take other factors into account that constrain the frame rate lower than this value, but it should not perform streaming at a higher frame rate than this value.</description>
    </param>
</struct>
```

#### Updating video streaming capabilites during ignition cycle

> SDL Core 7.1 adds the ability to update video streaming capabilities during the ignition cycle. This will allow uses cases which require dynamic resolution switching (Picture-in-Picture, preview, split screen, etc.)

Reviewer: ```suggestion
SDL Core 7.1 adds the ability for an application to update its video streaming capabilities during the ignition cycle. This will allow SDL to handle uses cases which require dynamic resolution switching (Picture-in-Picture, preview, split screen, etc.)
```

> - Added new parameter `additionalVideoStreamingCapabilites` to the `VideoStreamingCapabilitity` struct.

Reviewer: ```suggestion
- Added new parameter `additionalVideoStreamingCapabilities` to the `VideoStreamingCapability` struct.
```

```xml
<struct name="VideoStreamingCapability" since="4.5">
    <!-- Existing params -->
    <param name="additionalVideoStreamingCapabilities" type="VideoStreamingCapability" array="true" minvalue="1" maxvalue="100" mandatory="false" since="7.1">
    </param>
</struct>
```

> - Added new RPC notification `OnAppCapabilityUpdated` and related structs `AppCapability` and `AppCapabilityType`.

Reviewer: ```suggestion
- Added new RPC notification `OnAppCapabilityUpdated` which can be sent by an app, as well as related structs `AppCapability` and `AppCapabilityType`.
```

- The function `file_system::ConvertPathForURL` has been marked as deprecated and should be removed in the next major version change of SDL Core. Please make updates to remove all uses of the function.

### Deprecated Vehicle Data

- Vehicle Data parameter `externalTemperature` has been deprecated. Please make updates to use the new vehicle data struct `climateData`.

- Vehicle Data parameters `driverDoorAjar`, `passengerDoorAjar`, `rearLeftDoorAjar` and `rearRightDoorAjar` have been deprecated. Please make updates to use the new `doorStatuses` parameter.

## Additions

### Vehicle Data

- `BodyInformation` was expanded to replace `driverDoorAjar`, `passengerDoorAjar`, `rearLeftDoorAjar` and `rearRightDoorAjar` parameters.
- New vehicle data type: `climateData` to replace `externalTemperature` parameter.
- New vehicle data type: `seatOccupancy`.

It is not required to implement all vehicle data types. If a type is unsupported by your headunit, please be sure to respond to SDL Core with result `UNSUPPORTED_RESOURCE` if an unsupported request has been made.

### HMI UI Additions

#### Custom Playback rates for SetMediaClockTimer

> A media app now has the ability when setting the media playback timer and progress bar to specify a custom playback rate (ex. 125% speed).

Reviewer: ```suggestion
A media app now has the ability to specify a custom playback rate (ex. 125% speed) when setting the media playback timer and progress bar.
```
